File: tools/generate_leverage_data.py
#!/usr/bin/env python3
"""
Sentio Trading System - Leverage Data Generator

This script generates theoretical leverage data for TQQQ, SQQQ, and PSQ
based on the base QQQ data using accurate pricing models.

Usage:
    python3 generate_leverage_data.py [--input QQQ_RTH_NH.csv] [--output-dir data/equities/]
"""

import pandas as pd
import numpy as np
import argparse
import os
from pathlib import Path
from datetime import datetime, timedelta
import math

class LeverageDataGenerator:
    """Generate theoretical leverage data from base QQQ data"""
    
    def __init__(self, daily_decay_rate=0.0001, expense_ratio=0.0095):
        """
        Initialize the leverage data generator
        
        Args:
            daily_decay_rate: Daily rebalancing cost (default: 0.01%)
            expense_ratio: Annual expense ratio (default: 0.95%)
        """
        self.daily_decay_rate = daily_decay_rate
        self.expense_ratio = expense_ratio
        
        # Leverage specifications
        self.leverage_specs = {
            'TQQQ': {'factor': 3.0, 'inverse': False, 'description': '3x Long QQQ'},
            'SQQQ': {'factor': 3.0, 'inverse': True, 'description': '3x Short QQQ'},
            'PSQ': {'factor': 1.0, 'inverse': True, 'description': '1x Short QQQ'}
        }
    
    def calculate_leverage_price(self, prev_lev_price, qqq_daily_return, leverage_factor, is_inverse):
        """
        Calculate theoretical leverage price using daily return compounding
        
        CORRECTED MODEL: This now uses the accurate daily return compounding approach
        that correctly models the path-dependent nature of leveraged ETFs.
        
        Args:
            prev_lev_price: Previous leveraged ETF price
            qqq_daily_return: Daily return of base QQQ asset
            leverage_factor: Leverage multiplier (1.0 or 3.0)
            is_inverse: Whether this is an inverse instrument
            
        Returns:
            New theoretical leverage price
        """
        # Apply leverage factor to daily return
        leveraged_return = qqq_daily_return * leverage_factor
        
        # Apply inversion if necessary
        if is_inverse:
            leveraged_return = -leveraged_return
        
        # Apply daily costs (decay + expense ratio)
        daily_costs = self.daily_decay_rate + (self.expense_ratio / 252.0)
        
        # Calculate new price using compounding
        new_price = prev_lev_price * (1.0 + leveraged_return - daily_costs)
        
        return max(new_price, 0.01)  # Ensure positive price
    
    def generate_leverage_series(self, qqq_data, symbol):
        """
        Generate leverage series for a specific symbol using corrected daily return compounding
        
        CORRECTED MODEL: This now uses the accurate daily return compounding approach
        that correctly models the path-dependent nature of leveraged ETFs.
        
        Args:
            qqq_data: DataFrame with QQQ OHLCV data
            symbol: Target symbol (TQQQ, SQQQ, PSQ)
            
        Returns:
            DataFrame with leverage OHLCV data
        """
        if symbol not in self.leverage_specs:
            raise ValueError(f"Unsupported leverage symbol: {symbol}")
        
        spec = self.leverage_specs[symbol]
        leverage_factor = spec['factor']
        is_inverse = spec['inverse']
        
        print(f"Generating {symbol} data ({spec['description']}) using corrected daily return compounding...")
        
        # Create output dataframe
        leverage_data = qqq_data.copy()
        
        # Initialize starting price for leveraged ETF
        starting_price = 100.0
        if is_inverse:
            starting_price = 50.0  # Inverse ETFs typically start lower
        
        prev_lev_price = starting_price
        
        # Calculate theoretical prices using daily return compounding with progress tracking
        total_rows = len(qqq_data)
        print(f"Processing {total_rows:,} rows for {symbol}...")
        
        # Process in chunks for better performance
        chunk_size = 10000
        for chunk_start in range(1, total_rows, chunk_size):
            chunk_end = min(chunk_start + chunk_size, total_rows)
            
            # Show progress
            progress_pct = (chunk_start / total_rows) * 100
            print(f"  Progress: {progress_pct:.1f}% ({chunk_start:,}/{total_rows:,} rows)")
            
            # Process chunk
            for i in range(chunk_start, chunk_end):
                prev_qqq_close = qqq_data.iloc[i-1]['close']
                curr_qqq_close = qqq_data.iloc[i]['close']
                
                # Calculate daily return of base QQQ
                qqq_daily_return = (curr_qqq_close / prev_qqq_close) - 1.0
                
                # Calculate new leveraged price using compounding
                curr_lev_price = self.calculate_leverage_price(
                    prev_lev_price, qqq_daily_return, leverage_factor, is_inverse
                )
                
                # Calculate OHLC based on intraday movements
                curr_qqq_open = qqq_data.iloc[i]['open']
                curr_qqq_high = qqq_data.iloc[i]['high']
                curr_qqq_low = qqq_data.iloc[i]['low']
                
                # Calculate intraday movement ratios
                open_ratio = (curr_qqq_open - prev_qqq_close) / prev_qqq_close
                high_ratio = (curr_qqq_high - prev_qqq_close) / prev_qqq_close
                low_ratio = (curr_qqq_low - prev_qqq_close) / prev_qqq_close
                
                # Apply leverage factor to intraday movements
                lev_open_ratio = open_ratio * leverage_factor * (-1 if is_inverse else 1)
                lev_high_ratio = high_ratio * leverage_factor * (-1 if is_inverse else 1)
                lev_low_ratio = low_ratio * leverage_factor * (-1 if is_inverse else 1)
                
                # Calculate leveraged OHLC
                lev_open = prev_lev_price * (1.0 + lev_open_ratio)
                lev_high = prev_lev_price * (1.0 + lev_high_ratio)
                lev_low = prev_lev_price * (1.0 + lev_low_ratio)
                
                # For inverse ETFs, high and low are swapped
                if is_inverse:
                    lev_high, lev_low = lev_low, lev_high
                
                # Ensure OHLC relationships
                leverage_data.iloc[i, leverage_data.columns.get_loc('open')] = max(lev_open, 0.01)
                leverage_data.iloc[i, leverage_data.columns.get_loc('high')] = max(lev_high, lev_open, curr_lev_price, 0.01)
                leverage_data.iloc[i, leverage_data.columns.get_loc('low')] = max(min(lev_low, lev_open, curr_lev_price), 0.01)
                leverage_data.iloc[i, leverage_data.columns.get_loc('close')] = curr_lev_price
                
                # Adjust volume (typically lower for leverage instruments)
                volume_scaling = 0.3 if leverage_factor == 3.0 else 0.5
                leverage_data.iloc[i, leverage_data.columns.get_loc('volume')] = int(qqq_data.iloc[i]['volume'] * volume_scaling)
                
                prev_lev_price = curr_lev_price
        
        print(f"Generated {len(leverage_data)} bars for {symbol} using corrected daily return compounding")
        return leverage_data
    
    def generate_all_leverage_data(self, qqq_file_path, output_dir, max_rows=None):
        """
        Generate all leverage data files
        
        Args:
            qqq_file_path: Path to QQQ base data file
            output_dir: Directory to save leverage data files
            max_rows: Maximum number of rows to process (for testing)
        """
        print(f"Loading QQQ data from: {qqq_file_path}")
        
        # Load QQQ data
        try:
            qqq_data = pd.read_csv(qqq_file_path)
            original_rows = len(qqq_data)
            
            # Limit rows if specified (for testing)
            if max_rows and max_rows < original_rows:
                qqq_data = qqq_data.head(max_rows)
                print(f"⚠️  Limited to {max_rows:,} rows for testing (original: {original_rows:,})")
            
            print(f"Loaded {len(qqq_data):,} QQQ bars")
        except Exception as e:
            print(f"Error loading QQQ data: {e}")
            return False
        
        # Ensure output directory exists
        output_path = Path(output_dir)
        output_path.mkdir(parents=True, exist_ok=True)
        
        # Generate leverage data for each symbol
        for symbol in self.leverage_specs.keys():
            try:
                leverage_data = self.generate_leverage_series(qqq_data, symbol)
                
                # Save to file
                output_file = output_path / f"{symbol}_RTH_NH.csv"
                leverage_data.to_csv(output_file, index=False)
                
                print(f"Saved {symbol} data to: {output_file}")
                
                # Display sample data
                print(f"Sample {symbol} data:")
                print(leverage_data.head(3))
                print()
                
            except Exception as e:
                print(f"Error generating {symbol} data: {e}")
                return False
        
        print("✅ All leverage data files generated successfully!")
        return True

def main():
    """Main function"""
    parser = argparse.ArgumentParser(description='Generate leverage data for Sentio Trading System')
    parser.add_argument('--input', default='data/equities/QQQ_RTH_NH.csv',
                       help='Path to QQQ base data file')
    parser.add_argument('--output-dir', default='data/equities/',
                       help='Directory to save leverage data files')
    parser.add_argument('--daily-decay', type=float, default=0.0001,
                       help='Daily decay rate for leverage instruments')
    parser.add_argument('--expense-ratio', type=float, default=0.0095,
                       help='Annual expense ratio')
    parser.add_argument('--max-rows', type=int, help='Maximum number of rows to process (for testing)')
    
    args = parser.parse_args()
    
    # Check if input file exists
    if not os.path.exists(args.input):
        print(f"❌ Error: Input file not found: {args.input}")
        print("Please ensure QQQ_RTH_NH.csv exists in the data/equities/ directory")
        return 1
    
    # Create generator
    generator = LeverageDataGenerator(
        daily_decay_rate=args.daily_decay,
        expense_ratio=args.expense_ratio
    )
    
    # Generate leverage data
    success = generator.generate_all_leverage_data(args.input, args.output_dir, args.max_rows)
    
    if success:
        print("\n🎯 Leverage Data Generation Complete!")
        print("Generated files:")
        print("  - TQQQ_RTH_NH.csv (3x Long QQQ)")
        print("  - SQQQ_RTH_NH.csv (3x Short QQQ)")
        print("  - PSQ_RTH_NH.csv (1x Short QQQ)")
        print("\nThese files can now be used for leverage trading in sentio_cli trade command.")
        return 0
    else:
        print("❌ Leverage data generation failed!")
        return 1

if __name__ == "__main__":
    exit(main())

File: tools/generate_leverage_data.cpp
// =============================================================================
// Executable: generate_leverage_data
// Purpose: Generates theoretical data for TQQQ, SQQQ, PSQ from a base QQQ file.
//          Fulfills requirements from Section 3 of the Leverage Trading doc.
//
// Core Architecture & Design:
// This tool implements the corrected daily return compounding model that
// accurately captures the path-dependent nature of leveraged ETFs through
// proper daily rebalancing simulation rather than simple price multiplication.
//
// Key Features:
// - Daily return compounding: Models actual ETF rebalancing mechanism
// - Path dependency: Captures volatility decay and compounding effects
// - Realistic pricing: Generates data that matches actual ETF behavior
// - Accurate OHLC: Maintains proper intraday relationships
// - Volume scaling: Realistic volume adjustments for leverage instruments
//
// Usage:
//   generate_leverage_data --input data/equities/QQQ_RTH_NH.csv --output-dir data/equities/
// =============================================================================

#include "common/utils.h"
#include "common/types.h"
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <cmath>
#include <fstream>
#include <iomanip>
#include <stdexcept>

/// Leverage specification for data generation
struct LeverageSpec {
    double factor;      /// Leverage multiplier (1.0 or 3.0)
    bool is_inverse;     /// Whether instrument is inverse
    std::string description; /// Human-readable description
};

/// Calculate leveraged bar using corrected daily return compounding model
/// 
/// This implementation uses the accurate daily return compounding approach
/// that correctly models the path-dependent nature of leveraged ETFs.
/// 
/// @param qqq_bar Current QQQ bar
/// @param prev_qqq_bar Previous QQQ bar
/// @param prev_lev_bar Previous leveraged bar
/// @param spec Leverage specification
/// @param daily_decay Daily decay rate
/// @return Generated leveraged bar
sentio::Bar calculate_leveraged_bar(const sentio::Bar& qqq_bar, 
                                   const sentio::Bar& prev_qqq_bar,
                                   const sentio::Bar& prev_lev_bar, 
                                   const LeverageSpec& spec,
                                   double daily_decay) {
    sentio::Bar lev_bar = qqq_bar; // Copy timestamps and symbol initially
    
    // 1. Calculate daily return of base asset (QQQ)
    double qqq_return = (qqq_bar.close / prev_qqq_bar.close) - 1.0;
    
    // 2. Apply leverage factor to daily return
    double leveraged_return = qqq_return * spec.factor;
    
    // 3. Apply inversion if necessary
    if (spec.is_inverse) {
        leveraged_return = -leveraged_return;
    }
    
    // 4. Apply daily costs (decay)
    double total_daily_cost = daily_decay;
    
    // 5. Calculate new leveraged price using compounding
    lev_bar.close = prev_lev_bar.close * (1.0 + leveraged_return - total_daily_cost);
    
    // Ensure positive price
    lev_bar.close = std::max(0.01, lev_bar.close);
    
    // 6. Generate OHLC based on intraday movements
    double open_move = (qqq_bar.open - prev_qqq_bar.close) / prev_qqq_bar.close;
    lev_bar.open = prev_lev_bar.close * (1.0 + open_move * spec.factor * (spec.is_inverse ? -1.0 : 1.0));
    
    // Calculate high and low based on intraday movements
    double high_move = (qqq_bar.high - qqq_bar.open) / qqq_bar.open;
    double low_move = (qqq_bar.low - qqq_bar.open) / qqq_bar.open;
    
    if (spec.is_inverse) {
        // For inverse ETFs, high becomes low and low becomes high
        lev_bar.high = lev_bar.open * (1.0 + low_move * spec.factor * -1.0);
        lev_bar.low = lev_bar.open * (1.0 + high_move * spec.factor * -1.0);
    } else {
        lev_bar.high = lev_bar.open * (1.0 + high_move * spec.factor);
        lev_bar.low = lev_bar.open * (1.0 + low_move * spec.factor);
    }
    
    // Ensure OHLC integrity
    double max_oc = std::max(lev_bar.open, lev_bar.close);
    double min_oc = std::min(lev_bar.open, lev_bar.close);
    lev_bar.high = std::max(lev_bar.high, max_oc);
    lev_bar.low = std::min(lev_bar.low, min_oc);
    
    // Ensure all prices are positive
    lev_bar.open = std::max(0.01, lev_bar.open);
    lev_bar.high = std::max(0.01, lev_bar.high);
    lev_bar.low = std::max(0.01, lev_bar.low);
    
    // Model volume as a fraction (typically lower for leverage instruments)
    double volume_scaling = (spec.factor == 3.0) ? 0.3 : 0.5;
    lev_bar.volume = static_cast<int>(qqq_bar.volume * volume_scaling);
    
    return lev_bar;
}

/// Write leverage data to CSV file
/// @param bars Vector of leverage bars
/// @param symbol Symbol name
/// @param output_path Output file path
/// @return true if successful
bool write_leverage_csv(const std::vector<sentio::Bar>& bars, 
                       const std::string& symbol, 
                       const std::string& output_path) {
    std::ofstream out(output_path);
    if (!out.is_open()) {
        std::cerr << "ERROR: Cannot open output file: " << output_path << std::endl;
        return false;
    }
    
    // Write CSV header
    out << "ts_utc,ts_nyt_epoch,open,high,low,close,volume\n";
    out << std::fixed << std::setprecision(2);
    
    // Write data rows
    for (const auto& bar : bars) {
        out << sentio::utils::ms_to_timestamp(bar.timestamp_ms) << ","
            << bar.timestamp_ms / 1000 << ","
            << bar.open << "," << bar.high << "," << bar.low << "," << bar.close << ","
            << static_cast<long>(bar.volume) << "\n";
    }
    
    return true;
}

int main(int argc, char** argv) {
    // Parse command line arguments
    const std::string input_path = sentio::utils::get_arg(argc, argv, "--input", "data/equities/QQQ_RTH_NH.csv");
    const std::string output_dir = sentio::utils::get_arg(argc, argv, "--output-dir", "data/equities/");
    const double daily_decay = std::stod(sentio::utils::get_arg(argc, argv, "--decay", "0.0001"));
    const int max_rows = std::stoi(sentio::utils::get_arg(argc, argv, "--max-rows", "0"));
    
    std::cout << "=============================================================================" << std::endl;
    std::cout << "Sentio Leverage Data Generator - Corrected Daily Return Compounding Model" << std::endl;
    std::cout << "=============================================================================" << std::endl;
    std::cout << "Input file: " << input_path << std::endl;
    std::cout << "Output directory: " << output_dir << std::endl;
    std::cout << "Daily decay rate: " << daily_decay << std::endl;
    if (max_rows > 0) {
        std::cout << "Max rows (testing): " << max_rows << std::endl;
    }
    std::cout << std::endl;
    
    // Load base QQQ data
    std::cout << "📊 Loading base QQQ data from: " << input_path << std::endl;
    auto qqq_bars = sentio::utils::read_csv_data(input_path);
    if (qqq_bars.empty()) {
        std::cerr << "❌ ERROR: Failed to load QQQ data from " << input_path << std::endl;
        return 1;
    }
    
    // Limit rows if specified (for testing)
    if (max_rows > 0 && max_rows < static_cast<int>(qqq_bars.size())) {
        qqq_bars.resize(max_rows);
        std::cout << "⚠️  Limited to " << max_rows << " rows for testing" << std::endl;
    }
    
    std::cout << "✅ Loaded " << qqq_bars.size() << " QQQ bars" << std::endl;
    std::cout << std::endl;
    
    // Define leverage specifications
    std::map<std::string, LeverageSpec> specs = {
        {"TQQQ", {3.0, false, "3x Long QQQ"}},
        {"SQQQ", {3.0, true, "3x Short QQQ"}},
        {"PSQ",  {1.0, true, "1x Short QQQ"}}
    };
    
    // Generate leverage data for each symbol
    for (const auto& pair : specs) {
        const std::string& symbol = pair.first;
        const LeverageSpec& spec = pair.second;
        
        std::cout << "🔧 Generating " << symbol << " data (" << spec.description << ")..." << std::endl;
        std::cout << "   Using corrected daily return compounding model" << std::endl;
        
        std::vector<sentio::Bar> leverage_bars;
        leverage_bars.reserve(qqq_bars.size());
        
        // Initialize the first bar
        sentio::Bar first_lev_bar = qqq_bars[0];
        first_lev_bar.symbol = symbol;
        
        // Set starting price based on instrument type
        double starting_price = spec.is_inverse ? 50.0 : 100.0;
        first_lev_bar.open = first_lev_bar.high = first_lev_bar.low = first_lev_bar.close = starting_price;
        
        leverage_bars.push_back(first_lev_bar);
        
        // Process remaining bars using daily return compounding
        for (size_t i = 1; i < qqq_bars.size(); ++i) {
            leverage_bars.push_back(
                calculate_leveraged_bar(qqq_bars[i], qqq_bars[i-1], leverage_bars.back(), spec, daily_decay)
            );
            leverage_bars.back().symbol = symbol;
            
            // Show progress for large datasets
            if (qqq_bars.size() > 10000 && i % 10000 == 0) {
                double progress = (static_cast<double>(i) / qqq_bars.size()) * 100.0;
                std::cout << "   Progress: " << std::fixed << std::setprecision(1) << progress << "%" << std::endl;
            }
        }
        
        // Write to CSV
        std::string out_path = output_dir + symbol + "_RTH_NH.csv";
        if (write_leverage_csv(leverage_bars, symbol, out_path)) {
            std::cout << "✅ Successfully saved " << leverage_bars.size() << " bars to " << out_path << std::endl;
            
            // Show sample data
            if (!leverage_bars.empty()) {
                const auto& sample = leverage_bars[0];
                std::cout << "   Sample: Open=" << sample.open << ", High=" << sample.high 
                         << ", Low=" << sample.low << ", Close=" << sample.close << std::endl;
            }
        } else {
            std::cerr << "❌ Failed to save " << symbol << " data" << std::endl;
            return 1;
        }
        
        std::cout << std::endl;
    }
    
    std::cout << "🎯 Leverage Data Generation Complete!" << std::endl;
    std::cout << "Generated files:" << std::endl;
    std::cout << "  - TQQQ_RTH_NH.csv (3x Long QQQ)" << std::endl;
    std::cout << "  - SQQQ_RTH_NH.csv (3x Short QQQ)" << std::endl;
    std::cout << "  - PSQ_RTH_NH.csv (1x Short QQQ)" << std::endl;
    std::cout << std::endl;
    std::cout << "These files can now be used for leverage trading in sentio_cli trade command." << std::endl;
    
    return 0;
}
